pkg/provider/kubernetes/gateway/kubernetes.go

```diff
@@ -356,167 +407,367 @@ func (p *Provider) fillGatewayConf(client Client, gateway *v1alpha1.Gateway, con
+			if routeRule.ForwardTo == nil {
+				continue
+			}
+
+			wrrService, subServices, err := loadTCPServices(client, gateway.Namespace, routeRule.ForwardTo)
+			if err != nil {
+				// update "ResolvedRefs" status true with "DroppedRoutes" reason
+				conditions = append(conditions, metav1.Condition{
+					Type:               string(v1alpha1.ListenerConditionResolvedRefs),
+					Status:             metav1.ConditionFalse,
+					LastTransitionTime: metav1.Now(),
+					Reason:             string(v1alpha1.ListenerReasonDegradedRoutes),
+					Message:            fmt.Sprintf("Cannot load service from %s %s/%s : %v", listener.Routes.Kind, gateway.Namespace, tcpRoute.Name, err),
+				})
+
+				// TODO update the RouteStatus condition / deduplicate conditions on listener
+				continue
+			}
+
+			for svcName, svc := range subServices {
+				conf.TCP.Services[svcName] = svc
+			}
+
+			serviceName := provider.Normalize(routerKey + "-wrr")
+			conf.TCP.Services[serviceName] = wrrService
+
+			router.Service = serviceName
+
+			routerKey = provider.Normalize(routerKey)
+			conf.TCP.Routers[routerKey] = &router
+		}
+	}
+
+	return conditions
+}
 
-					conf.HTTP.Routers[routerKey] = &router
+func gatewayTLSRouteToTCPConf(ep string, listener v1alpha1.Listener, gateway *v1alpha1.Gateway, client Client, conf *dynamic.Configuration) []metav1.Condition {
+	// TODO: support RouteNamespaces
+	tlsRoutes, err := client.GetTLSRoutes(gateway.Namespace, labels.SelectorFromSet(listener.Routes.Selector.MatchLabels))
+	if err != nil {
+		// update "ResolvedRefs" status true with "InvalidRoutesRef" reason
+		return []metav1.Condition{{
+			Type:               string(v1alpha1.ListenerConditionResolvedRefs),
+			Status:             metav1.ConditionFalse,
+			LastTransitionTime: metav1.Now(),
+			Reason:             string(v1alpha1.ListenerReasonInvalidRoutesRef),
+			Message:            fmt.Sprintf("Cannot fetch %ss for namespace %q and matchLabels %v", listener.Routes.Kind, gateway.Namespace, listener.Routes.Selector.MatchLabels),
+		}}
+	}
+
+	var conditions []metav1.Condition
+	for _, tlsRoute := range tlsRoutes {
+		for _, routeRule := range tlsRoute.Spec.Rules {
+			rule, err := hostSNIRule(routeRule)
+			if err != nil {
+				// update "ResolvedRefs" status true with "DroppedRoutes" reason
+				conditions = append(conditions, metav1.Condition{
+					Type:               string(v1alpha1.ListenerConditionResolvedRefs),
+					Status:             metav1.ConditionFalse,
+					LastTransitionTime: metav1.Now(),
+					Reason:             string(v1alpha1.ListenerReasonDegradedRoutes),
+					Message:            fmt.Sprintf("Skipping %s %s: cannot make route's SNI match: %v", listener.Routes.Kind, tlsRoute.Name, err),
+				})
+				// TODO update the RouteStatus condition / deduplicate conditions on listener
+				continue
+			}
+
+			router := dynamic.TCPRouter{
+				Rule:        rule,
+				EntryPoints: []string{ep},
+			}
+
+			if listener.TLS != nil {
```

Comment:
Since we are in gatewayTLSRouteToTCPConf , shouldn't this always be the case? Shouldn't it be an error if listener.TLS == nil ? (not sure, asking just in case).